### DISCUSSION TITLE
Ability to dynamically reduce center layout padding when there is less space

### DISCUSSION THREAD

**Discussion by tealsnow** [zed-industries/zed#24612] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

I use the centered layout most of the time, but when I open something like the assistant panel, it ends up squishing the editor view too small (as seen in the attached screenshot). I can change the left and right padding statically using the `centered_layout` settings. But this makes the padding too small without the panel. I find my self toggling the center layout far too often than I would like.

I propose that there should be a configuration for minimum and maximum padding in `centered_layout`, let's say I set the minimum to 0 and the max to the default of 0.2. In the event that there is not enough space to display enough of the editor, it will resize within the range configured. The target size (columns) could perhaps be determined by the `wrap_guides` setting or a new one in `centered_layout`. Thus, in the event that I open the assistant panel that the editor will expand to a comfortable size, then when I close it, the padding will increase again.

For example:
```json
"centered_layout": {
  "min_left_padding": 0,
  "max_left_padding": 0.2,
  "min_right_padding": 0,
  "max_right_padding": 0.2,
  "target_column_width": 80
}
```

### If applicable, add mockups / screenshots to help present your vision of the feature

![2024-10-30_09-35-35](https://github.com/user-attachments/assets/588dac7c-4cf3-4b21-8a6c-80071600c8b9)

**Comment by Serdan** [comment]
I installed Zed today. This is the first "wtf" I've run into. Defining centered layout by statically assigned padding is completely backwards. What I want is an editor panel with a width I consider comfortable. It's the _width_ that should remain consistent, regardless of window size or other panels. 
If having the panel be off-center is something people want for their "centered" editor, the appropriate solution would be an `offset`.

**Comment by Morssel** [comment]
Agreed. First issue I have run into that I really didn't like how the Zed editor backed me into a corner.

I want to use centered layout, but then it squishes my editor file window (like in OP's screenshot) when I open a panel like the agent chat or terminal dock on the left/right side of the editor.

Fix could be as easy as an option under centered_layout like "disable_when_panel_open": true. The layout logic could then treat an open left/right dock the same way it already treats a second editor pane today (which correctly drops centering).

**Comment by Serdan** [comment]
I was so bothered by this that I vibe-coded what I consider the correct design. https://github.com/Serdan/zed
New setting is:
```json
"centered_layout": {
  "max_width": 1255,
},
```

- Can drag pane border
- Actions `increase centered width` and `decrease centered width`
- Supports centered split pane as well, for ultra-widescreen users.

Fork also has better handling of hidden files because that was another thing that was annoying me (`toggle show hidden`).

**Comment by marcoms** [comment]
I agree that specifying padding percentages isn't ideal as it doesn't account for resizing the window. It would be great to use the editor preferred line length as the initial size (e.g. `min(windowWidth, 80 characters)`) and let the user drag-resize the column if they want to.
